src/zarr/core/array.py

```diff
@@ -344,7 +347,9 @@ async def _create_v3(
 
         array = cls(metadata=metadata, store_path=store_path)
         await array._save_metadata(metadata, ensure_parents=True)
-        return array
+        # type inference is inconsistent here and seems to conclude
```

Comment:
I thought the issue might be from `AsyncArray.metadata` being set to `metadata_parsed = parse_array_metadata(metadata)`, which doesn't know that `parse_array_metadata(ArrayV2Metadata)` should be `ArrayV2Metadata`. I thought that adding an overload would fix that:

```python
@overload
def parse_array_metadata(data: ArrayV2Metadata) -> ArrayV2Metadata: ...


@overload
def parse_array_metadata(data: ArrayV3Metadata) -> ArrayV3Metadata: ...


@overload
def parse_array_metadata(data: dict[str, JSON]) -> ArrayV2Metadata | ArrayV3Metadata: ...


def parse_array_metadata(
    data: ArrayV2Metadata | ArrayV3Metadata | dict[str, JSON],
) -> ArrayV2Metadata | ArrayV3Metadata:
 ```

But no luck. I'll take a longer look later.

Reply:
Here's a simple example that works:

<details>

```python
from __future__ import annotations

import dataclasses
from typing import Generic, Literal, TypeVar, cast, overload, reveal_type


class A: ...

class B: ...

T = TypeVar("T", A, B)

@dataclasses.dataclass(frozen=True)
class Array(Generic[T]):
    version: T

    def __init__(self, version: A | B) -> None:
        object.__setattr__(self, "version", version)

    @overload
    @classmethod
    def create(cls, version: Literal[2]) -> Array[A]: ...

    @overload
    @classmethod
    def create(cls, version: Literal[3]) -> Array[B]: ...
 
    @overload
    @classmethod
    def create(cls, version: Literal[2, 3]) -> Array[A] | Array[B]: ...

    @classmethod
    def create(cls, version: Literal[2, 3]) -> Array[A] | Array[B]:
        if version == 2:
            return cls.create_2()
        elif version == 3:
            return cls.create_3()
        else:
            raise

    @classmethod
    def create_2(cls) -> Array[A]:
        return Array(version=A())

    @classmethod
    def create_3(cls) -> Array[B]:
        return Array(version=B())

reveal_type(Array(A()))
reveal_type(Array(B()))

reveal_type(Array.create_2())
reveal_type(Array.create_3())

reveal_type(Array.create(version=2))
reveal_type(Array.create(version=3))
```

</details>

That reveals the expected types everywhere:

```
check.py:49: note: Revealed type is "check.Array[check.A]"
check.py:50: note: Revealed type is "check.Array[check.A]"
check.py:52: note: Revealed type is "check.Array[check.A]"
check.py:53: note: Revealed type is "check.Array[check.B]"
check.py:55: note: Revealed type is "check.Array[check.A]"
check.py:56: note: Revealed type is "check.Array[check.B]"
Success: no issues found in 1 source file
```

I've tried making similar changes on your branch, but haven't gotten it working yet. I think the things that might matter are:

1. We might need overloads for `AsyncArray.create` to ensure that `AsyncArray.create(..., zarr_format=3)` returns an `AsyncArray[v3]`. likewise for 2
2. `AsyncArray.__init__` also accepts a `dict` for `metadata`. The actual value set is the output of `parse_arary_metadata`. We need to ensure that the specialized type isn't lost (which the overload *should* do, but maybe I messed that up).

Reply:
I'm not sure what I messed up earlier, but these overloads on `create` seemed to do the trick:

```diff
diff --git a/src/zarr/core/array.py b/src/zarr/core/array.py
index 96431dec..5699935f 100644
--- a/src/zarr/core/array.py
+++ b/src/zarr/core/array.py
@@ -4,7 +4,7 @@ import json
 from asyncio import gather
 from dataclasses import dataclass, field, replace
 from logging import getLogger
-from typing import TYPE_CHECKING, Any, Generic, Literal, TypeVar, cast
+from typing import TYPE_CHECKING, Any, Generic, Literal, TypeVar, cast, overload
 
 import numpy as np
 import numpy.typing as npt
@@ -191,6 +191,72 @@ class AsyncArray(Generic[TArrayMeta]):
         object.__setattr__(self, "order", order_parsed)
         object.__setattr__(self, "codec_pipeline", create_codec_pipeline(metadata=metadata_parsed))
 
+    @overload
+    @classmethod
+    async def create(
+        cls,
+        store: StoreLike,
+        *,
+        # v2 and v3
+        shape: ShapeLike,
+        dtype: npt.DTypeLike,
+        zarr_format: Literal[2],
+        fill_value: Any | None = None,
+        attributes: dict[str, JSON] | None = None,
+        # v3 only
+        chunk_shape: ChunkCoords | None = None,
+        chunk_key_encoding: (
+            ChunkKeyEncoding
+            | tuple[Literal["default"], Literal[".", "/"]]
+            | tuple[Literal["v2"], Literal[".", "/"]]
+            | None
+        ) = None,
+        codecs: Iterable[Codec | dict[str, JSON]] | None = None,
+        dimension_names: Iterable[str] | None = None,
+        # v2 only
+        chunks: ShapeLike | None = None,
+        dimension_separator: Literal[".", "/"] | None = None,
+        order: Literal["C", "F"] | None = None,
+        filters: list[dict[str, JSON]] | None = None,
+        compressor: dict[str, JSON] | None = None,
+        # runtime
+        exists_ok: bool = False,
+        data: npt.ArrayLike | None = None,
+    ) -> AsyncArray[ArrayV2Metadata]:...
+
+    @overload
+    @classmethod
+    async def create(
+        cls,
+        store: StoreLike,
+        *,
+        # v2 and v3
+        shape: ShapeLike,
+        dtype: npt.DTypeLike,
+        zarr_format: Literal[3],
+        fill_value: Any | None = None,
+        attributes: dict[str, JSON] | None = None,
+        # v3 only
+        chunk_shape: ChunkCoords | None = None,
+        chunk_key_encoding: (
+            ChunkKeyEncoding
+            | tuple[Literal["default"], Literal[".", "/"]]
+            | tuple[Literal["v2"], Literal[".", "/"]]
+            | None
+        ) = None,
+        codecs: Iterable[Codec | dict[str, JSON]] | None = None,
+        dimension_names: Iterable[str] | None = None,
+        # v2 only
+        chunks: ShapeLike | None = None,
+        dimension_separator: Literal[".", "/"] | None = None,
+        order: Literal["C", "F"] | None = None,
+        filters: list[dict[str, JSON]] | None = None,
+        compressor: dict[str, JSON] | None = None,
+        # runtime
+        exists_ok: bool = False,
+        data: npt.ArrayLike | None = None,
+    ) -> AsyncArray[ArrayV3Metadata]:...
+
     @classmethod
     async def create(
         cls,
@@ -349,7 +415,7 @@ class AsyncArray(Generic[TArrayMeta]):
         await array._save_metadata(metadata, ensure_parents=True)
         # type inference is inconsistent here and seems to conclude
         # that array has type Array[ArrayV2Metadata]
-        return array  # type: ignore[return-value]
+        return array
 
     @classmethod
     async def _create_v2(
@@ -388,7 +454,7 @@ class AsyncArray(Generic[TArrayMeta]):
         )
         array = cls(metadata=metadata, store_path=store_path)
         await array._save_metadata(metadata, ensure_parents=True)
-        return array  # type: ignore[return-value]
+        return array
 
     @classmethod
     def from_dict(
```

Checking that here:

```python
file: check2.py
from typing import reveal_type

import numpy as np

from zarr.core.array import AsyncArray
from zarr.storage.common import StorePath
from zarr.storage.memory import MemoryStore


async def main() -> None:
    store = StorePath(MemoryStore(), "/")
    r2 = await AsyncArray._create_v2(store, shape=(0,), dtype=np.dtype("float64"), chunks=(0,))
    reveal_type(r2)  # v2

    r3 = await AsyncArray._create_v3(store, shape=(0,), dtype=np.dtype("float64"), chunk_shape=(0,))
    reveal_type(r3)  # v3

    rr2 = await AsyncArray.create(store, shape=(0,), dtype=np.dtype("float64"), chunks=(0,), zarr_format=2)
    reveal_type(rr2)  # v2

    rr3 = await AsyncArray.create(store, shape=(0,), dtype=np.dtype("float64"), chunks=(0,), zarr_format=3)
    reveal_type(rr3)  # v3
```

gives

```
❯ mypy check2.py
check2.py:13: note: Revealed type is "zarr.core.array.AsyncArray[zarr.core.metadata.v2.ArrayV2Metadata]"
check2.py:16: note: Revealed type is "zarr.core.array.AsyncArray[zarr.core.metadata.v3.ArrayV3Metadata]"
check2.py:19: note: Revealed type is "zarr.core.array.AsyncArray[zarr.core.metadata.v2.ArrayV2Metadata]"
check2.py:22: note: Revealed type is "zarr.core.array.AsyncArray[zarr.core.metadata.v3.ArrayV3Metadata]"
Success: no issues found in 1 source file
```